src/main/java/uk/gov/homeoffice/digital/sas/balancecalculator/BalanceCalculator.java

```diff
@@ -0,0 +1,232 @@
+  static final String ACCRUALS_NOT_FOUND =
+      "No Accrual records found for tenantId {0} and personId {1} between {2} and {3}";
+  static final String MISSING_ACCRUAL =
+      "Accrual missing for tenantId {0}, personId {1}, accrual type {2} and date {3}";
+  static final String ACCRUALS_MAP_EMPTY = "Accruals Map must contain at least one entry!";
+
+  private final RestClient restClient;
+  private final List<AccrualModule> accrualModules;
+
+  @Autowired
+  public BalanceCalculator(RestClient restClient, List<AccrualModule> accrualModules) {
+    this.restClient = restClient;
+    this.accrualModules = accrualModules;
+  }
+
+  public List<Accrual> calculate(TimeEntry timeEntry) {
+
+    String timeEntryId = timeEntry.getId();
+    String tenantId = timeEntry.getTenantId();
+    String personId = timeEntry.getOwnerId();
+    ZonedDateTime timeEntryStart = timeEntry.getActualStartTime();
+    ZonedDateTime timeEntryEnd = timeEntry.getActualEndTime();
+    LocalDate timeEntryStartDate = timeEntryStart.toLocalDate();
+    LocalDate timeEntryEndDate = timeEntryEnd.toLocalDate();
+
+    // Get agreement applicable to the end date of the time entry (in case the time entry spans two
+    // agreements
+    Agreement applicableAgreement =
+        getAgreementApplicableToTimeEntryEndDate(tenantId, personId, timeEntryEndDate);
+    if (applicableAgreement == null) {
+      log.warn(MessageFormat.format(AGREEMENT_NOT_FOUND, tenantId, personId, timeEntryEndDate));
+      return List.of();
+    }
+
+    // Get accruals of all types between the day just before the time entry and the end date of the
+    // latest applicable agreement
+    Map<AccrualType, TreeMap<LocalDate, Accrual>> allAccruals =
+        getAccrualsBetweenDates(tenantId, personId,
+            timeEntryStartDate.minusDays(1), applicableAgreement.getEndDate());
+
+    if (isEmpty(allAccruals)) {
+      log.warn(MessageFormat.format(ACCRUALS_NOT_FOUND, tenantId, personId,
+          timeEntryStartDate.minusDays(1), applicableAgreement.getEndDate()));
+      return List.of();
+    }
+
+    Map<LocalDate, Range<ZonedDateTime>> dateRangeMap = splitOverDays(timeEntryStart, timeEntryEnd);
+
+    for (var entry : dateRangeMap.entrySet()) {
+      for (var module : accrualModules) {
+        LocalDate referenceDate = entry.getKey();
+        ZonedDateTime startTime = entry.getValue().lowerEndpoint();
+        ZonedDateTime endTime = entry.getValue().upperEndpoint();
+        AccrualType accrualType = module.getAccrualType();
+        TreeMap<LocalDate, Accrual> accruals = allAccruals.get(accrualType);
+
+        Accrual accrual = accruals.get(referenceDate);
+
+        if (accrual == null) {
+          log.error(MessageFormat.format(
+              MISSING_ACCRUAL, tenantId, personId, accrualType, referenceDate));
+          return List.of();
+        }
+
+        BigDecimal shiftContribution = module.calculateShiftContribution(startTime, endTime);
```

Comment:
I think the order of some of the looping isn't correct here and possible may not come to light until we start other modules. For example shit counts, if the module is only passed the start and end time for the split day it may not count a shift in the correct place. e.g. 11pm-7am should count a shift on the first day but the first day would only get the data for the first hour.
I think the looping should look more like
for each accrual type {
  const contribs[] = module.getContributions(timeEntry); // [{ date, value }]
  for each contribs[] {
    updateAccrualContribution;
    cascadeCumulativeTotal
  }
}

Can we quickly discuss, it might be that this goes into the split day ticket

Reply:
Had a chance to catch up with @taomoh and both agreed that this was useful insight into how it may look in the future but didn't require change for now as it wouldn't event annualised target hours. @taomoh is looking at the advantage of including this in the next ticket vs waiting until we work on impacted accruals such as shift counts.